src/query-builder/json-path-builder.ts

```diff
@@ -0,0 +1,276 @@
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('nicknames', '->>$').at('#-1').as('last_nickname')
+   * )
+   * ```
+   *
+   * The generated SQL (SQLite):
+   *
+   * ```sql
+   * select "nicknames"->>'$[#-1]' as `last_nickname` from `person`
+   * ```
+   */
+  at<
+    I extends any[] extends O ? number | 'last' | `#-${number}` : never,
+    O2 = null | NonNullable<NonNullable<O>[keyof NonNullable<O> & number]>
+  >(
+    index: `${I}` extends `${any}.${any}` | `#--${any}` ? never : I
+  ): TraversedJSONPathBuilder<S, O2> {
+    return this.#createBuilderWithPathLeg('ArrayLocation', index)
+  }
+
+  /**
+   * Access a property of a JSON object.
+   *
+   * If a field is optional, the resulting type will be nullable.
+   *
+   * See also {@link at} to access elements of JSON arrays.
+   *
+   * ### Examples
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('address', '->').key('city').as('city')
+   * )
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * select "address"->'city' as "city" from "person"
+   * ```
+   *
+   * Going deeper:
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('profile', '->$').key('website').key('url').as('website_url')
+   * )
+   * ```
+   *
+   * The generated SQL (MySQL):
+   *
+   * ```sql
+   * select `profile`->'$.website.url' as `website_url` from `person`
+   * ```
+   *
+   * Combined with {@link at}:
+   *
+   * ```ts
+   * db.selectFrom('person').select(eb =>
+   *   eb.ref('profile', '->').key('addresses').at(0).key('city').as('city')
+   * )
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * select "profile"->'addresses'->0->'city' as "city" from "person"
+   * ```
+   */
+  key<
+    K extends any[] extends O
+      ? never
+      : O extends object
+      ? keyof NonNullable<O> & string
+      : never,
+    O2 = undefined extends O
+      ? null | NonNullable<NonNullable<O>[K]>
+      : null extends O
+      ? null | NonNullable<NonNullable<O>[K]>
+      : NonNullable<O>[K]
+  >(key: K): TraversedJSONPathBuilder<S, O2> {
+    return this.#createBuilderWithPathLeg('Member', key)
+  }
+
+  #createBuilderWithPathLeg(
+    legType: JSONPathLegType,
+    value: string | number
+  ): TraversedJSONPathBuilder<any, any> {
+    const legNode = JSONPathLegNode.create(
+      legType,
+      RawNode.createWithSql(String(value))
+    )
+
+    return new TraversedJSONPathBuilder(
+      ReferenceNode.is(this.#node)
+        ? ReferenceNode.cloneWithJSONPath(
+            this.#node,
+            JSONPathReferenceNode.clone(
+              this.#node.jsonPath!,
+              JSONPathNode.cloneWithLeg(this.#node.jsonPath!.jsonPath, legNode)
+            )
+          )
+        : JSONPathNode.cloneWithLeg(this.#node, legNode)
+    )
+  }
+}
+
+export class TraversedJSONPathBuilder<S, O>
+  extends JSONPathBuilder<S, O>
+  implements Expression<O>
+{
+  readonly #node: ReferenceNode | JSONPathNode
```

Comment:
Same here. Always a `ReferenceNode`?

Reply:
Simplified for now @ 4ab875d1587b027ee1090fa6fa9c5439a85fbc02.